src/Lachain.Networking/NetworkManagerBase.cs

```diff
@@ -145,22 +171,33 @@ private void _HandleMessage(object sender, byte[] buffer)
             }
         }
 
-        private Action<IMessage> SendTo(ClientWorker peer)
+        private Action<IMessage> SendTo(ClientWorker peer, ulong requestId)
         {
             return x =>
             {
                 Logger.LogTrace($"Sending {x.GetType()} to {peer.PeerPublicKey.ToHex()}");
                 NetworkMessage msg = x switch
                 {
-                    PingReply pingReply => new NetworkMessage {PingReply = pingReply},
-                    SyncBlocksReply syncBlockReply => new NetworkMessage {SyncBlocksReply = syncBlockReply},
-                    SyncPoolReply syncPoolReply => new NetworkMessage {SyncPoolReply = syncPoolReply},
-                    GetPeersReply getPeersReply => new NetworkMessage {GetPeersReply = getPeersReply},
+                    PingReply pingReply => new NetworkMessage { RequestId = requestId, PingReply = pingReply },
+                    SyncBlocksReply syncBlockReply => new NetworkMessage
+                        { RequestId = requestId, SyncBlocksReply = syncBlockReply },
+                    SyncPoolReply syncPoolReply => new NetworkMessage
+                        { RequestId = requestId, SyncPoolReply = syncPoolReply },
+                    GetPeersReply getPeersReply => new NetworkMessage
+                        { RequestId = requestId, GetPeersReply = getPeersReply },
                     _ => throw new InvalidOperationException()
                 };
-                // we should never reply with priority, requests can be spammed
-                peer.AddMsgToQueue(msg, NetworkMessagePriority.ReplyMessage);
-            };
+
+                if (_requestIdentifierSet.Contains(new RequestIdentifier(requestId, peer.PeerPublicKey)))
```

Comment:
This is wrong. We should check this when we receive a reply. Here we are sending reply.